### PR TITLE
[REVIEW] Python cuda kernel example

### PR DESCRIPTION
**Pull request by jdye64**
This closes: #4829

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by jdye64** [comment]
I just noticed that as well. Looks like I just mistakenly opened the PR against 0.14 instead of 0.15 as intended. I'll fix it

**Comment by kkraus14** [comment]
Haven't dug deeply into the code yet, but it feels like we should add a docs page giving a high level overview of the pieces of this example and how they fit together. Also, is `examples` where we want things like this to live? cc @rapidsai/ops

File: examples/.gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+python-kernel-wrapper/cython/*.cpp
+python-kernel-wrapper/cython/*.so
+python-kernel-wrapper/cython/record.txt
+
+# Ignore data files download by the script
+python-kernel-wrapper/data/weather
```
Comment (harrism) [inline examples/.gitignore:6]:
```suggestion
python-kernel-wrapper/data/weather

```

File: examples/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# cuDF Examples
+This directory contains examples of cuDF usage. Some examples are simple while others are more involved. Each example is held within a subdirectory. Each example will have its own README to explain intended use, assist with building, and running.
```
Comment (harrism) [inline examples/README.md:2]: There's only one example, so how can you say some are simple while others are more involved?

Comment (harrism) [inline examples/README.md:2]:
```suggestion
This directory contains examples of cuDF usage.
```

File: examples/python-kernel-wrapper/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Python Kernel Wrapper
+
+## Overview
+This examples aims to demonstrate how its possible to share cudf dataframes between Python and CUDA Kernels. This is useful if either an existing CUDA codebase is present to solve business logic or if teams would like the ability to handle certain tasks in Python and others in CUDA.
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:4]:
```suggestion
This example demonstrates how to share cudf dataframes between Python and custom CUDA kernels. This is useful for performing custom CUDA-accelerated business logic on cuDF dataframes and handling certain tasks in Python and others in CUDA.
```

File: examples/python-kernel-wrapper/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+Since cudf uses libcudf, dataframes that are created in Python are already present on the GPU memory and available to the C++ codebase. This makes it fairly straightforward to write a CUDA kernel for working with a dataframe column. In fact this is how libcudf processes dataframes in CUDA kernels within cudf itself, the only difference in this example is we are writing CUDA kernels that are not in the cudf codebase. The term User Defined Function (UDF) could be loosely used to describe what this example is demonstrating.
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:6]:
```suggestion
Dataframes that are created in Python cuDF are already present in GPU memory and accessible to CUDA code. This makes it straightforward to write a CUDA kernel to work with a dataframe columns. In fact this is how libcudf processes dataframes in CUDA kernels; the only difference in this example is that we invoke CUDA kernels that exist outside the cuDF code base. The term User Defined Function (UDF) could be loosely used to describe what this example is demonstrating.
```
I would maybe drop the UDF sentence. UDFs tend to follow a well-defined interface and are defined outside cuDF but are applied to data frames or columns *by* cuDF. So this is loose enough to possibly cause confusion.

File: examples/python-kernel-wrapper/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+A cython ```kernel_wrapper``` implementation has been created in this example to further make sharing the dataframes between Python and your CUDA kernel "UDF" easier. This wrapper allows for Python users to seamless invoke those "UDF" CUDA kernels with a single function call and also provides a clear place for implementing the C++ "glue code".
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:8]:
```suggestion
This example provides a Cython `kernel_wrapper` implementation to make sharing the dataframes between Python and our custom CUDA kernel easier. This wrapper allows Python users to seamlessly invoke those CUDA kernels with a single function call and also provides a clear place to implement the C++ "glue code".
```

File: examples/python-kernel-wrapper/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+Now that we understand the motivation and how it will work lets define the problem for this example.
+
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:11]:
```suggestion
```

File: examples/python-kernel-wrapper/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+The goal is to write a CUDA kernel that accepts a rainfall (PRCP) data column and convert those values to inches. The original data is stored as 1/10th of a mm so the kernel must convert those values to inches. The dataframe will be read from a local csv file and created using Python. Python will then invoke the CUDA kernel doing the mm->inches conversion with the ```kernel_wrapper``` and the dataframe object. Once the kernel has finished its execution the dataframe values will be changed on the Python side as well so those can be seen by simple output with ```df.head()```.
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:12]:
```suggestion
The example CUDA kernel accepts a data column (PRCP) containing rainfall values stored as 1/10th of a mm and converts those values to inches. The dataframe is read from a local CSV file using Python. Python then invokes the CUDA mm->inches conversion kernel via the Cython `kernel_wrapper`, passing it the dataframe object. The converted data can then be accessed from Python, e.g. using `df.head()`.
```

File: examples/python-kernel-wrapper/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+This is similar to what is being done in an existing [weather notebook](https://github.com/rapidsai/notebooks-contrib/blob/branch-0.14/intermediate_notebooks/examples/weather.ipynb) so if any confusion arises that is a good place to reference for general understanding.
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:14]:
```suggestion
This is similar to an existing [weather notebook](https://github.com/rapidsai/notebooks-contrib/blob/branch-0.14/intermediate_notebooks/examples/weather.ipynb), which provides a reference for understanding the implementation. 
```

File: examples/python-kernel-wrapper/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+2. [CUDA](https://developer.nvidia.com/cuda-downloads) 10.0 > is installed and on the PATH.
+
+## Build Steps
+1. Clone the cudf repo if you haven't already ```git clone https://github.com/rapidsai/cudf.git && cd cudf/examples/python-kernel-wrapper```
+2. Create the ```cudf_ex``` conda environment. ```conda env create -f ./conda/cudf_ex.yml --name cudf_ex```
+3. Activate the ```cudf_ex``` sconda environment ```conda activate cudf_ex```
+3. Build the cython "kernel_wrapper" code. ```cd cython && python setup.py build install``` Notice the custom Kernel definitions are in ```cython/src/kernel.cu```. This is just for example and the build can be altered as needed.
+4. Download weather data. A convenience Python script has been provided here to make that easier for you. By default it will download years 2010-2020 weather data. That data is about 300MB per file so if you need to download less files you can change that in the script. The data will be downloaded to ./data/weather. ```python ./data/download_data.py```
+5. Run the Python example script. It expects an input of a single Weather year file. EX: ```python ./01_python_kernel_wrapper.py ./data/weather/2010.csv.gz```
```
Comment (harrism) [inline examples/python-kernel-wrapper/README.md:28]:
```suggestion
## Build Steps
1. Clone the cudf repo if you haven't already:
```
git clone https://github.com/rapidsai/cudf.git && cd cudf/examples/python-kernel-wrapper
```
2. Create the `cudf_ex` conda environment. 
```
conda env create -f ./conda/cudf_ex.yml --name cudf_ex
```
3. Activate the `cudf_ex` conda environment:
```
conda activate cudf_ex
```
3. Build the cython `kernel_wrapper` code: 
```
cd cython && python setup.py build install
```.
Notice the custom kernel definitions are in `cython/src/kernel.cu`. This is just for example and the build can be altered as needed.
4. Download weather data. A convenience Python script has been provided here to make that easier. By default it will download years 2010-2020 weather data. The data is about 300MB per file so if you need to download fewer files you can change that in the script. The data will be downloaded to `./data/weather`. 
```
python ./data/download_data.py
```
5. Run the Python example script. It expects an input of a single Weather year file. 
```
python ./01_python_kernel_wrapper.py ./data/weather/2010.csv.gz
```
```

Comment (harrism) [reply examples/python-kernel-wrapper/README.md:28]: Github screwed up my suggestion due to the ```

File: examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  // Invoke the Kernel to convert tenth_mm -> inches
+  kernel_tenth_mm_to_inches<<<(mtv.num_rows()+255)/256, 256>>>(*mutable_device_column, mtv.num_rows());
+  cudaDeviceSynchronize();
+}
+
+CudfWrapper::~CudfWrapper() {
+  // It is important to note that CudfWrapper does not own the underlying Dataframe 
+  // object and that will be freed by the Python/Cython layer later.
+}
```
Comment (harrism) [inline examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu:50]:
```suggestion
}

```

File: examples/python-kernel-wrapper/cython/src/kernel.cu [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+static const float mm_to_inches = 0.0393701;
+
+__global__ void kernel_tenth_mm_to_inches(cudf::mutable_column_device_view column, int64_t num_rows)
+{
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < num_rows) {
+        column.element<int64_t>(i) = column.element<int64_t>(i) * (1/10) * mm_to_inches;
+    }
+}
```
Comment (harrism) [inline examples/python-kernel-wrapper/cython/src/kernel.cu:13]: Couple of things. 
 - Use constexpr
 - size of columns is `cudf::size_type` which is 32-bit. But you don't need that parameter anyway, since it's part of the device_view.

```suggestion
static constexpr float mm_to_inches = 0.0393701;

__global__ void kernel_tenth_mm_to_inches(cudf::mutable_column_device_view column)
{
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if (i < column.size()) {
        column.element<int64_t>(i) = column.element<int64_t>(i) * (1/10) * mm_to_inches;
    }
}
```

Technically this should also use a grid-stride loop rather than an if() because if we every change size() to be a 64-bit value, it will be possible to have more rows than you can process with a single 1D CUDA grid.

To be honest, best practice would be to use `thrust::transform` with a lambda to implement this rather than a kernel...

File: examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+void CudfWrapper::tenth_mm_to_inches(int column_index) {
+
+  // Example of showing num_columns and num_rows only for potential debugging
+  printf("kernel_wrapper.cu # of columns: %lu\n", mtv.num_columns());
+  printf("kernel_wrapper.cu # of rows: %lu\n", mtv.num_rows());
+
+  // Raw loop to print out column dtypes for example sake only, not required,
+  for( cudf::size_type c = 0; c < mtv.num_columns(); c++ ) {
+    printf("%d type=%d, ptr=%p\n", c, (int)mtv.column(c).type().id(), mtv.column(c).data<char>() );
+  }
```
Comment (harrism) [inline examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu:37]:
```suggestion
  std::for_each( mtv.cbegin(), mtv.cend(), [](auto c) { 
    printf("%d type=%d, ptr=%p\n", c, static_cast<int>(c.type().id()), c.data<char>() );
  });
```

File: examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <kernel.cu>
+#include <kernel_wrapper.hh>
+#include <assert.h>
+#include <iostream>
```
Comment (harrism) [inline examples/python-kernel-wrapper/cython/src/kernel_wrapper.cu:20]: You include iostream but you use printf. Either switch printf to cout, or include `<cstdio>` instead.

File: examples/python-kernel-wrapper/cython/src/kernel_wrapper.hh [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#include <string>
```
Comment (harrism) [inline examples/python-kernel-wrapper/cython/src/kernel_wrapper.hh:1]: Can you use .hpp for the extension for consistency with libcudf style?

File: examples/README.md (hunk 2 above)
Comment (jdye64) [reply examples/README.md:2]: You are right. I was thinking ahead about other examples I had planned to add. I will reword it.

Comment (jdye64) [reply examples/README.md:2]: Actually I'll just remove that part for now.

Comment (jdye64) [reply examples/README.md:2]: ahh I see you thought the same as I mentioned ^^ =)

File: examples/python-kernel-wrapper/README.md (hunk 9 above)
Comment (jdye64) [reply examples/python-kernel-wrapper/README.md:28]: I think I can tell what you are suggesting here but might want to check again after my next commit since with the github format issue its hard to follow.

File: examples/python-kernel-wrapper/cython/setup.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            "and try again."
+        )
+    CUDA_HOME = os.path.dirname(os.path.dirname(path_to_cuda_gdb))
+
+cuda_include_dir = os.path.join(CUDA_HOME, "include")
+cuda_lib_dir = os.path.join(CUDA_HOME, "lib64")
+
+
+# Include the ability to compile using both GCC and NVCC
+def compile_cuda(self):
```
[the remark this answers is not in the document]
Comment (jdye64) [reply examples/python-kernel-wrapper/cython/setup.py:28]: I wouldn't say it common that is for sure. However the person asking for an example like this was so I just continued that way. I can change the CUDA C++ build to CMake if you want?

**Comment by jdye64** [comment]
@kkraus14 I just chose `examples` at random. I'm open to putting it anywhere

File: examples/python-kernel-wrapper/cython/kernel.pyx [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+        void tenth_mm_to_inches(int column_index)
+
+cdef class CudfWrapper:
+    cdef C_CudfWrapper* gdf
+
+    def __cinit__(self, Table t):
+        self.gdf = new C_CudfWrapper(t.mutable_view())
+
+    def tenth_mm_to_inches(self, col_index):
+        self.gdf.tenth_mm_to_inches(col_index)
```
Comment (kkraus14) [inline examples/python-kernel-wrapper/cython/kernel.pyx:18]: Is there a reason the example is based around making a C++ class instead of just a free function? I think the example would be more concise and easier to follow if `tenth_mm_to_inches` was just a free function taking in a `cudf::(mutable_)table_view` as input and returning a `cudf::(mutable_)table_view` or nothing if it just modifies the table inplace.

File: examples/python-kernel-wrapper/cython/setup.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        raise OSError(
+            "Could not locate CUDA. "
+            "Please set the environment variable "
+            "CUDA_HOME to the path to the CUDA installation "
+            "and try again."
+        )
+    CUDA_HOME = os.path.dirname(os.path.dirname(path_to_cuda_gdb))
+
+cuda_include_dir = os.path.join(CUDA_HOME, "include")
+cuda_lib_dir = os.path.join(CUDA_HOME, "lib64")
```
Comment (kkraus14) [inline examples/python-kernel-wrapper/cython/setup.py:24]: I don't think you need to find / include CUDA here since the Cython you've written doesn't directly depend on CUDA in any way. (I could be mistaken though in just cimporting cudf things could trigger the dependency).

File: examples/python-kernel-wrapper/cython/setup.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (kkraus14) [reply examples/python-kernel-wrapper/cython/setup.py:28]: Agreed that the CUDA / C++ here you likely wouldn't want to try to compile as a Python extension.

File: examples/python-kernel-wrapper/README.md [hunk 18]
```diff
@@ -0,0 +1,26 @@
+# Python Kernel Wrapper
+
+## Overview
+This example demonstrates how to share cudf dataframes between Python and custom CUDA kernels. This is useful for performing custom CUDA-accelerated business logic on cuDF dataframes and handling certain tasks in Python and others in CUDA.
+
+Dataframes that are created in Python cuDF are already present in GPU memory and accessible to CUDA code. This makes it straightforward to write a CUDA kernel to work with a dataframe columns. In fact this is how libcudf processes dataframes in CUDA kernels; the only difference in this example is that we invoke CUDA kernels that exist outside the cuDF code base. The term User Defined Function (UDF) could be loosely used to describe what this example is demonstrating.
```
Comment (kkraus14) [inline examples/python-kernel-wrapper/README.md:6]:
```suggestion
Dataframes that are created in Python cuDF are already present in GPU memory and accessible to CUDA code. This makes it straightforward to write a CUDA kernel to work with a dataframe's columns. In fact this is how libcudf processes dataframes in CUDA kernels; the only difference in this example is that we invoke CUDA kernels that exist outside the cuDF code base. The term User Defined Function (UDF) could be loosely used to describe what this example is demonstrating.
```

File: examples/python-kernel-wrapper/conda/cudf_ex.yml [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+  - conda-forge
+  - defaults
+dependencies:
+  - cudf=0.14.*
+  - rmm=0.14.*
+  - cmake>=3.12
+  - cmake_setuptools>=0.1.3
+  - python>=3.6,<3.8
+  - cython>=0.29,<0.30
+
```
Comment (kkraus14) [inline examples/python-kernel-wrapper/conda/cudf_ex.yml:15]:
```suggestion

```

File: examples/python-kernel-wrapper/cython/kernel.pyx [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+cdef extern from "src/kernel_wrapper.hpp":
+    cdef cppclass C_CudfWrapper "CudfWrapper":
+        C_CudfWrapper(mutable_table_view tbl)
+        void tenth_mm_to_inches(int column_index)
+
+cdef class CudfWrapper:
+    cdef C_CudfWrapper* gdf
+
+    def __cinit__(self, Table t):
+        self.gdf = new C_CudfWrapper(t.mutable_view())
```
Comment (kkraus14) [inline examples/python-kernel-wrapper/cython/kernel.pyx:15]: You're currently leaking this.

File: examples/python-kernel-wrapper/cython/kernel.pyx (hunk 16 above)
Comment (harrism) [reply examples/python-kernel-wrapper/cython/kernel.pyx:18]: This is a good point. The cython will definitely be simpler, as will the C++/CUDA.

File: examples/python-kernel-wrapper/cython/setup.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jdye64) [reply examples/python-kernel-wrapper/cython/setup.py:28]: ok, I'm going to change all of this to cmake now which should cover most of the other existing suggestions as well.

[the remark this answers is not in the document]
Comment (jdye64) [reply examples/python-kernel-wrapper/cython/setup.py:28]: My opinion is adding a cmake 'find_package' module for cudf should be done before this happens. I have opened an issue for that [here](https://github.com/rapidsai/cudf/issues/5569)

**Comment by harrism** [comment]
@jdye64 is this waiting on cudf find_package or something else?

**Comment by jdye64** [comment]
> @jdye64 is this waiting on cudf find_package or something else?

Yes, Waiting on some upstream CMake changes from others.

**Comment by harrism** [comment]
@jdye64 what's the status of this now?

**Comment by jdye64** [comment]
> @jdye64 what's the status of this now?

I'm waiting on a strategy for libcudf to be packaged up in a portable manner, its dependencies that is. I assume that will be something like a CPM package.

**Comment by jdye64** [comment]
@harrism how would you feel about moving `examples`, which would include this PR, to another repo? Something like `rapids/rapids-examples`? I ask since there are quite a few other examples I would like to add and it seems like that might possible be a more clean place to add things of that nature. Thoughts?

Also @kkraus14?

**Comment by harrism** [comment]
If the examples depend on and demonstrate multiple RAPIDS library APIs it definitely should be another repo. But if they are just cuDF examples, I don't feel strongly either way.

**Comment by jrhemstad** [comment]
What's the status of this PR? Moving to 0.19.

**Comment by caryr35** [comment]
Per Slack from Jeremy on 5/5/21 [](https://nvidia.slack.com/team/UL3R4GC2J )"It can just be deleted. I thought we had already. We moved those examples to another repo `rapids-examples` so it isn't needed anymore".  I am moving forward with closing this PR
